### DISCUSSION TITLE
Useage with urql client (with access token in the headers)

### DISCUSSION THREAD

**Discussion by oldo** [nextauthjs/next-auth#1806] [Help] [answered]
I have a situation that I am sure many people must of come across before. I'm using urql as a graphql client and I'd like to create the client with the headers set with the auth token.

The file that initialises the urql client looks like so:

```javascript
import { createClient } from 'urql';

const client = createClient({
  url: process.env.NEXT_PUBLIC_API_URL,
  fetchOptions: () => {
    return {
      headers: {}, // I'd like to set Auth header here...
    };
  },
});

export default client;
```

And then this is added to `_app.tsx`  via the `GraphqlProvider` as follows:

```javascript
import type { AppProps } from 'next/app';
import { Provider as AuthProvider } from 'next-auth/client';
import { Provider as GraphqlProvider } from 'urql';
import client from '../graphql/client';

function App({ Component, pageProps }: AppProps) {
  const { session } = pageProps;

  return (
    <GraphqlProvider value={client}>
      <AuthProvider session={session}>
        <Component {...pageProps} />
      </AuthProvider>
    </GraphqlProvider>
  );
}

export default App;
```

Nothing fancy at this stage.

Now my issue is that I want to be able to access the access token to add an Authorization header in the urql client, however because this bit of code exists outside of a React component I can't use the `useSession` hook provided by next-auth.

Can anyone suggest the best way to approach this issue? Thanks.

**Comment by balazsorban44** [comment, marked as answer]
the urql initialization should be a function, that for example takes an access token parameter.

Here is how we do it at work:
```js
import * as React from "react"
import GQLClient from "./GQLClient"
import { useErrorHandler } from "react-error-boundary"
import { useSession } from "next-auth/client"

/**
 * Get GraphQL Client in browser environments (frontend).
 *
 * If the user has an active session, it will add an accessToken to all requests
 */
export function useClient(options?: RequestInit) {
  const [session] = useSession()
  const token = session?.accessToken
  const handleError = useErrorHandler()
  return React.useMemo(() => {
    const client = GQLClient(
      process.env.NEXT_PUBLIC_CONTENT_API,
      {
        ...options,
        headers: { Authorization: `Bearer ${token}`, ...options?.headers },
      },
      handleError
    )
    return client
  }, [handleError, options, token])
}
```

Change GQLClient with urql

**Reply by oldo** [reply]
Oooo, that sounds promising! I have tried to implement and still haven't been able to get the token to be set in the header.
Here is what I've got:


`/graphql/client.ts`
```javascript
import { createClient } from '@urql/core';
import { useSession } from 'next-auth/client';
import * as React from 'react';

/**
 * Get GraphQL Client in browser environments (frontend).
 *
 * If the user has an active session, it will add an accessToken to all requests
 */
const useClient = (options?: RequestInit) => {
  const [session] = useSession();
  const token = session?.accessToken;
  // const handleError = useErrorHandler();

  return React.useMemo(() => {
    const client = createClient({
      url: process.env.NEXT_PUBLIC_API_URL,
      fetchOptions: () => {
        return {
          headers: {
            Authorization: token ? `Bearer ${token}` : '',
            ...(options?.headers ? options.headers : {}),
          },
        };
      },
    });

    return client;
  }, [options, token]);
};

export default useClient;
```

`/pages/_app.tsx`
```javascript
import type { AppProps } from 'next/app';
import { Provider as AuthProvider } from 'next-auth/client';
import { Provider as GraphqlProvider } from 'urql';
import useClient from '../graphql/client';

import '../styles/index.css';

function App({ Component, pageProps }: AppProps) {
  const { session } = pageProps;
  const client = useClient(); // <----------------------------------- using the hook here

  return (
    <GraphqlProvider value={client}>
      <AuthProvider session={session}>
        <Component {...pageProps} />
      </AuthProvider>
    </GraphqlProvider>
  );
}

export default App;
```

`/pages/index.tsx`
```javascript
import { signIn, signOut, useSession } from 'next-auth/client';
import React from 'react';
import { useQuery } from 'urql';

interface HomeProps {}

const Home: React.FC<HomeProps> = ({}) => {
  const [session, loading] = useSession();

  const QUERY = `
  query {
    users {
      id
      name
      email
    }
  }
  `;

  const request = useQuery({ query: QUERY, pause: !!session });
  console.log(request);
  
  return (
    <div>
      {!session && (
        <>
          Not signed in <br />
          <button onClick={() => signIn()}>Sign in</button>
        </>
      )}

      {session && (
        <>
          Signed in as {session.user.email} <br />
          <button onClick={() => signOut()}>Sign out</button>
        </>
      )}
    </div>
  );
};

export default Home;
```

You can see that I am delaying the `useQuery` hook until `session` has been initialised, however I'm still not getting the auth token in the header.

Do you have any suggestions? Thanks so much for your help! :-)

**Reply by balazsorban44** [reply]
Do you actually have an `accessToken` in your session? remember to explicitly send it to the client with the `jwt` and `session` callbacks.https://next-auth.js.org/configuration/callbacks

Other than that, I wouldn't call `useClient` directly in `_app`, or at least not above `AuthProvider`. Because `useSession` won't be able to utilize the caching of the session, meaning you will contact the /api/auth/session endpoint more than necessary. Make sure that `AuthProvider` is above your `GraphqlProvider`, and don't call `useClient` above `AuthProvider` either. You will likely need an extra layer.

**Reply by oldo** [reply]
Spot on! Thank you @balazsorban44 
So I moved the `AuthProvider` to be the most parent component in `_app.tsx` and then created a new component as follows:

`/graphql/provider.tsx`
```javascript
import React from 'react';
import { Provider } from 'urql';
import useClient from './client';

interface GraphqlProviderProps {}

const GraphqlProvider: React.FC<GraphqlProviderProps> = ({ children }) => {
  const client = useClient();

  return <Provider value={client}>{children}</Provider>;
};

export default GraphqlProvider;
```
And then use this as child to `AuthProvider` in `_app.tsx`
```javascript
import { Provider as AuthProvider } from 'next-auth/client';
import type { AppProps } from 'next/app';
import GraphqlProvider from '../graphql/provider';
import '../styles/index.css';

function App({ Component, pageProps }: AppProps) {
  const { session } = pageProps;

  return (
    <AuthProvider session={session}>
      <GraphqlProvider>
        <Component {...pageProps} />
      </GraphqlProvider>
    </AuthProvider>
  );
}

export default App;
```
And it works!
Thank you so much for your assistance :-)

**Reply by balazsorban44** [reply]
Awesome! It looks great! Happy to help. 😊 Renamed the discussion, so it's more searchable to others in the future. 😉

**Comment by liho00** [comment]
how about axios interceptor instead of graphql client, axios interceptor does not support pause... im struggling setting Authorization Bearer token header in axios headers... pls guide me... @balazsorban44 @oldo thanks in advance!

**Reply by oldo** [reply]
@liho00 I'm not sure what you mean by axios interceptor does not support pause. Do you mean that interceptor can not handle async logic? If so, it is actually possible:

```
client.interceptors.request.use(async config => {
  const token = await getToken();

  if (token && config.headers) {
    config.headers.authorization = `JWT ${token}`;
  }

  return config;
});
```

Note that this is not next-auth code, I'm just copying from a different codebase as a demonstration of async interceptor that will await the return of some data to attach auth header.

**Reply by liho00** [reply]
> @liho00 I'm not sure what you mean by axios interceptor does not support pause. Do you mean that interceptor can not handle async logic? If so, it is actually possible:
> 
> ```
> client.interceptors.request.use(async config => {
>   const token = await getToken();
> 
>   if (token && config.headers) {
>     config.headers.authorization = `JWT ${token}`;
>   }
> 
>   return config;
> });
> ```
> 
> Note that this is not next-auth code, I'm just copying from a different codebase as a demonstration of async interceptor that will await the return of some data to attach auth header.

This is not a proper or practical way, because this implementation will call getToken every API calls.
next auth getToken API call session endpoint at client side, if await getToken, it will slow down the whole application, if the webpage require call multiple api at the same time.

the ideal way or expected behavior should be get the access token then set in to axios common headers ONCE only.

**Comment by skwowet** [comment]
Thanks for the code snippet, helped me a lot to in initial set up @oldo 

I would like to add `ssrCache` in above setup, could you please share how?
